### ISSUE TITLE
implement DLM

### ISSUE THREAD

**Issue by erwanp** [radis/radis#37]
# Line Density Map

(Implementation status: https://github.com/radis/radis/pull/38) 

## Description

@dcmvdbekerom suggested to use a Line Density Map with lineshape templates: about a dozen typical lineshapes can be calculated instead of millions.  
(see initial discussion on [Gitter](https://gitter.im/radis-radiation/community?at=5d34b76ac8988013473b69c6) ) 

@dcmvdbekerom  wrote a comparison example ([CO2_test.zip](https://gitter.im/radis-radiation/community?at=5d3cea08f0ff3e2bba7d0377)) with ~57k lines. Using the DLM is >1000x faster than a naive implementation where all lines are convolved. 

I updated the script to compare the output with Radis directly. Updated example can be found in [CO2_test.py](https://gist.github.com/erwanp/4f7dd563947a5e03c849ee5900257d34#file-co2_test) using a reduced set of 2k lines. 

Results on my laptop below (v0.9.19 commit 5de0c68f ) : 

![image](https://user-images.githubusercontent.com/16088743/62947273-8568b080-bde2-11e9-9a66-4af40c9bd08b.png)

- DLM about 10,000x faster than the naive (reference) case
- DLM about 50x faster than Radis. 

Output spectra are very similar. The main difference is that @dcmvdbekerom  example does not include the partition functions, therefore spectra were normalized (differences should be observed for lines corresponding to the 2nd isotope whose partition function is twice that of the 1st ... however these lines are rare on this spectral range).

Radis being ~ 250 faster than the reference case is mainly because 1. we use a direct Voigt approximation (Whiting formula, executed in Fortran) and 2. the lineshape is calculated on a reduced spectral range rather than on an array the same size as the full spectral range.
The implementation of the DLM will probably negate the improvements in efficiency given by 1. & 2. However, as the head-on comparison shows, about a factor ~50 in performance improvement is expected. 

## Implementation 

Work is done on the branch [dlm](https://github.com/radis/radis/tree/dlm/radis), mainly in [lbl/broadening.py](https://github.com/radis/radis/blob/dlm/radis/lbl/broadening.py) 

Discussions on [![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/radis-radiation/community?at=5d34b76ac8988013473b69c6) 

Todo:
- [x] Reference case: use [CO2_test.py](https://gist.github.com/erwanp/4f7dd563947a5e03c849ee5900257d34#file-co2_test) mentioned previously
- [x] Aggregate lines on the 2 closest points of the spectral grid (as done by @dcmvdbekerom ), rather than on the closest (5de0c68f)
- [x] Same for weak lines calculations (f301683)
- [x] Rewrite existing calc_gaussian to work with numpy arrays instead of pandas dataframes (9fa0bb34)
- [x] Implement DLM in `_broaden_lines`  for equilibrium ([broadening.py](https://github.com/radis/radis/blob/dlm/radis/lbl/broadening.py)) with convolve (41cbc203)
- [x] Pass tests, see improvement in calculation times.
- [x] Implement for lines partly outside of spectral range 
- [x] Implement for nonequilibrium
- [x] Implement DLM with Voigt  (not really needed) 
Then, DLM v2:
- [x] Spectrum calculation with sum over DLM rather than sum over lines
- [x] **Calculate DLM with Fourier Transform**
- [ ] Automatically calculate appropriate DLM parameters. 
- [ ] Auto-warning if DLM not useful (low number of lines)

**Comment by erwanp** [comment]
@dcmvdbekerom: got a DLM implemented in the current architecture of RADIS. Version ee9783c yields about a x10 improvement factor (see test case below). 

![image](https://user-images.githubusercontent.com/16088743/63061156-812cb800-bef4-11e9-9242-db3e43f897f8.png)

Compared with your DLM example, this is still ~5 times slower. 
For larger databases (HITEMP 2250-2500 cm-1: 400k lines), RADIS gets to ~100s, DLM RADIS ~15s, while your DLM is still <1s. 

The difference is due to the current implementation of RADIS: the DLM is used as template of precomputed lineshapes. It makes the calculation extremely fast, but I still need to retrieve from memory one lineshape array per line. 
In your DLM, you distribute all the lines on the DLM before the convolution, therefore it remains blazing fast even with high number of lines. 

Implementing it exactly your way into Radis will require some deeper re-writing of the code, but so far the results are very promising.

**Comment by erwanp** [comment]
Implemented without loop over lines (f43b221). I now get the same calculation times as your DLM! 

![image](https://user-images.githubusercontent.com/16088743/63102896-48d0bc80-bf7c-11e9-951b-24ec752547b6.png)

**Comment by dcmvdbekerom** [comment]
Wow you made a lot of progress! I guess it makes sense that for a small number of lines a direct calculation is faster. 

I was reading your comments and I'm surprised that np.convolve is faster than FFT in some cases, because my understanding is that np.convolve uses 3 FFT's under the hood (although I may be mistaken). One thing you'll notice in my example is that if you use the FFT instead of np.convolve, you need to extend the spectral axis to prevent artifact at the edges. Ideally, it needs to be extended by a factor of 2, otherwise the (shoulders of the-) peaks at the edges will warp around the spectral axis.

Apart from the numpy.fft module, there are some modules that are supposed to be faster, notably pyFFTW. However, I did a quick comparison and for me np.fft was the fastest. Of course there are also GPU implementations which may improve performance (reikna.fft, scikits.cuda) but I have no experience with those.

**Comment by erwanp** [comment]
I just remembered that if you want to convolve a small kernel on a large spectrum you better do it in real space, and in the Fourier space if your two vectors are approximately the same length. But i've never tested it in practice. 

In the present case it seems that we're in the second category: on the example below, which is included in `test_broadening.py` in the latest commit, I work with a CO spectrum of only 221 lines : the DLM has only 12 templates (about 5,000 points), but it still requires 5s to convolve them on the spectral grid (about 150,000 points).  

```
Calculating line broadening (221 lines: expect ~ 0.1s on 1 CPU)
... Precomputed DLM lineshapes (12) in 0.0s
... Initialized vectors in 0.0s
... Get closest matching line & fraction in 0.0s
... Distribute lines over DLM 0.0s
... Convolve and sum on spectral range 4.9s
Calculated line broadening in 4.9s
process done in 5.0s
```
For comparison, in the previous implementation (no DLM) the only convolutions were done on 2 reduced-vectors (let's say 2x5000 points in the present case). Now the FT implementation becomes critical. 

I won't have time to work on that before Monday. If you want to give it a try, I prepared some of the code with NonImplementedError at the right places in the last commit (and a failing test in `test_broadening.py`) . But the file is quite a mess at the moment!
